wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java

```diff
@@ -0,0 +1,400 @@
+
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.data.schema.UnsupportedTypeException;
+import co.cask.cdap.api.service.http.HttpServiceRequest;
+import co.cask.cdap.api.service.http.HttpServiceResponder;
+import co.cask.wrangler.PropertyIds;
+import co.cask.wrangler.RequestExtractor;
+import co.cask.wrangler.ServiceUtils;
+import co.cask.wrangler.api.Row;
+import co.cask.wrangler.dataset.connections.Connection;
+import co.cask.wrangler.dataset.workspace.DataType;
+import co.cask.wrangler.dataset.workspace.WorkspaceDataset;
+import co.cask.wrangler.service.PluginConfiguration;
+import co.cask.wrangler.service.common.AbstractWranglerService;
+import co.cask.wrangler.service.connections.ConnectionType;
+import co.cask.wrangler.service.gcp.GCPUtils;
+import co.cask.wrangler.utils.ObjectSerDe;
+import com.google.cloud.ByteArray;
+import com.google.cloud.spanner.DatabaseId;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Options;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Type;
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.HttpURLConnection;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * Spanner data prep connection service
+ */
+public class SpannerService extends AbstractWranglerService {
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerService.class);
+  private static final String TABLE_NAME = "TableName";
+  // Spanner queries for listing tables and listing schema of table are documented at
+  // https://cloud.google.com/spanner/docs/information-schema
+  private static final Statement LIST_TABLES_STATEMENT =
+    Statement.of("SELECT t.table_name FROM " +
+                   "information_schema.tables AS t WHERE t.table_catalog = '' and t.table_schema = ''");
+
+  private static final Statement.Builder SCHEMA_STATEMENT_BUILDER =  Statement.newBuilder(
+    String.format("SELECT  t.column_name,t.spanner_type, t.is_nullable FROM information_schema.columns AS t WHERE " +
+                    "  t.table_catalog = ''  AND  t.table_schema = '' AND t.table_name = @%s", TABLE_NAME));
+  // Default Number of rows read from Spanner table by data-prep
+  private static final String DEFAULT_ROW_LIMIT = "1000";
+  private static final Gson GSON = new Gson();
+
+  /**
+   * Tests Spanner Connection.
+   */
+  @POST
+  @Path("/connections/spanner/test")
```

Comment:
is this the pattern followed by the other endpoints? If so, it's fine to follow the pattern, but this is not how paths should be. This would normally refer to a 'test' action on a connection with id 'spanner'.

Reply:
If this is how everything works, please open a jira.

Reply:
yes, there should be a single endpoint with connection-type however that is not the case currently. created JIRA https://issues.cask.co/browse/CDAP-14250